src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.CoreRT.cs

```diff
@@ -0,0 +1,207 @@
+                        flags |= dllImportSearchPathFlags;
+                    }
+
+                    ret = LocalLoadLibraryHelper(currLibNameVariation, flags, ref errorTracker);
+                    if (ret != IntPtr.Zero)
+                    {
+                        return ret;
+                    }
+                }
+                else if ((callingAssembly != null) && searchAssemblyDirectory)
+                {
+                    // Try to load the module alongside the assembly where the PInvoke was declared.
+                    // This only makes sense in dynamic scenarios (JIT/interpreter), so leaving this out for now.
+                }
+
+                ret = LocalLoadLibraryHelper(currLibNameVariation, dllImportSearchPathFlags, ref errorTracker);
+                if (ret != IntPtr.Zero)
+                {
+                    return ret;
+                }
+            }
+
+            return IntPtr.Zero;
+        }
+
+        private static IntPtr LoadFromPath(string libraryName, bool throwOnError)
+        {
+            LoadLibErrorTracker errorTracker = default;
+            IntPtr ret = LocalLoadLibraryHelper(libraryName, 0, ref errorTracker);
+            if (throwOnError && ret == IntPtr.Zero)
+            {
+                errorTracker.Throw(libraryName);
+            }
+
+            return ret;
+        }
+
+        private static IntPtr LocalLoadLibraryHelper(string libraryName, int flags, ref LoadLibErrorTracker errorTracker)
+        {
+#if PLATFORM_WINDOWS
+            IntPtr ret = Interop.mincore.LoadLibraryEx(libraryName, IntPtr.Zero, flags);
+            if (ret != IntPtr.Zero)
+            {
+                return ret;
+            }
+
+            int lastError = Marshal.GetLastWin32Error();
+            if (lastError != LoadLibErrorTracker.ERROR_INVALID_PARAMETER)
+            {
+                errorTracker.TrackErrorCode(lastError);
+            }
+
+            return ret;
+#else
+            IntPtr ret = IntPtr.Zero;
+            if (libraryName == null)
+            {
+                errorTracker.TrackErrorCode(LoadLibErrorTracker.ERROR_MOD_NOT_FOUND);
+            }
+            else if (libraryName == String.Empty)
+            {
+                errorTracker.TrackErrorCode(LoadLibErrorTracker.ERROR_INVALID_PARAMETER);
+            }
+            else
+            {
+                // TODO: FileDosToUnixPathA
+                ret = Interop.Sys.LoadLibrary(libraryName);
+                if (ret == IntPtr.Zero)
+                {
+                    errorTracker.TrackErrorCode(LoadLibErrorTracker.ERROR_MOD_NOT_FOUND);
+                }
+            }
+
+            return ret;
+#endif
+        }
+
+        private static void FreeLib(IntPtr handle)
+        {
+#if !PLATFORM_UNIX
+            bool result = Interop.mincore.FreeLibrary(handle);
+            if (!result)
+                throw new InvalidOperationException();
+#else
+            Interop.Sys.FreeLibrary(handle);
```

Comment:
I am not sure whether dlopen handles null gracefully.

Reply:
If you wanted to comment on the Interop.Sys.LoadLibrary call, that one already covers null and string.empty. FreeLibrary guards for null in the shared part of this method.

Reply:
I meant to comment on `Interop.Sys.FreeLibrary` / `dlclose`. I do not see it protecting against null anywhere and I have verified that calling `dlclose(NULL)` will crash:

```
jkotas@jkotas9:~$ cat test.c
#include <dlfcn.h>
#include <stdlib.h>

static void* handle = NULL;

int main()
{
    dlclose(handle);
}
jkotas@jkotas9:~$ gcc test.c -ldl
jkotas@jkotas9:~$ ./a.out
Segmentation fault (core dumped)
jkotas@jkotas9:~$
```